### PR TITLE
Improve ISSM interface

### PR DESCRIPTION
**Pull request by lostella**
*Issue #, if available:* addresses #1200 

*Description of changes:* Using custom ISSMs in DeepState (i.e. other than the default one for the given frequency), does not work because the time features from which coefficients are constructed always fall back to the default ones. This PR changes this, by coupling each ISSM with the time feature that is needed to derive its coefficients.

Tests are added for the ISSM classes, which were not there before.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/model/deepstate/test_issm.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        10,
+        issm.latent_dim(),
+        issm.latent_dim(),
+    )
+    assert innovation_coeff.shape == (2, 10, issm.latent_dim())
+
+
+def test_seasonality_issm_h():
+    issm = SeasonalityISSM(
+        num_seasons=12, time_feature=MonthOfYear(normalized=False)
```
Comment (lostella) [inline test/model/deepstate/test_issm.py:130]: The fact that `normalized=False` must be passed here is a bit annoying; as an alternative, we could either have alternative constructors that don't allow for normalization (eg `MonthOfYearIndex`) or just pass the feature *type* here, and construct the object internally.

Thoughts @rshyamsundar?

**Review by rshyamsundar — APPROVED** [review]

File: test/model/deepstate/test_issm.py (hunk 1 above)
Comment (rshyamsundar) [reply test/model/deepstate/test_issm.py]: yea.. passing only type and constructing the object internally as you did in the latest revision is better...

File: test/model/deepstate/test_issm.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+        )
+    )
+
+    emission_coeff, transition_coeff, innovation_coeff = issm.get_issm_coeff(
+        features
+    )
+
+    season_indices = [[0] * 10, [0] * 2 + [1] * 8]
+
+    for item in range(1, 2):
```
Comment (rshyamsundar) [inline test/model/deepstate/test_issm.py:187]: You are checking here for only one item?

File: test/model/deepstate/test_issm.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+        for time in range(2, 10):
+            season_indicator = mx.nd.one_hot(
+                mx.nd.array([season_indices[item][time]]), 12
```
Comment (rshyamsundar) [inline test/model/deepstate/test_issm.py:190]: Maybe I am missing something: `mx.nd.array([season_indices[item][time]])` is always `[1]` for this range of `item` and `time` values and you always create this one_hot array: `[[0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]`?

**Comment by rshyamsundar** [comment]
Also maybe can you run `benchmark_m4.py` with deep state estimator? Just to see there are no serious differences..

File: test/model/deepstate/test_issm.py (hunk 2 above)
Comment (lostella) [reply test/model/deepstate/test_issm.py:187]: right, that was for debugging, I'll revert it

File: test/model/deepstate/test_issm.py (hunk 3 above)
Comment (lostella) [reply test/model/deepstate/test_issm.py:190]: on `item=0` the season index is always `0`, on `item=1` the season index changes from `0` to `1` after two time steps

**Review by rshyamsundar — APPROVED** [review]

**Comment by lostella** [comment]
@rshyamsundar 

```
      dataset  mean_wQuantileLoss     sMAPE
0   m4_hourly            0.040880  0.134823
1    m4_daily            0.028031  0.036918
2   m4_weekly            0.046093  0.077881
3  m4_monthly            0.117584  0.142290
```

**Pull request merged**